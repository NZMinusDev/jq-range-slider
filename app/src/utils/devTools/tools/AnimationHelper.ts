/**
 *
 * @param drawCallBack
 * @param duration
 * @param timing - calculating the current animation state function
 * @example
 * animate(
 * (progress) => { // draw
 *    elem.style.width = progress * 100 + '%';
 *  }
 *  1000, // duration
 *  {timing: (timeFraction) => { // parabolic curve timing function
 *    return Math.pow(timeFraction, 2)
 *  }},
 * );
 *
 * // timing functions(but you can pass your own):
 * function circ(timeFraction) { return 1 - Math.sin(Math.acos(timeFraction)); }
 * function shotFromABow(x, timeFraction) { return Math.pow(timeFraction, 2) * ((x + 1) * timeFraction - x) }
 * function bounce(timeFraction) {
 *   for (let a = 0, b = 1; 1; a += b, b /= 2) {
 *     if (timeFraction >= (7 - 4 * a) / 11) {
 *       return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2)
 *     }
 *   }
 * }
 * function elastic(x, timeFraction) {
 *   return Math.pow(2, 10 * (timeFraction - 1)) * Math.cos(20 * Math.PI * x / 3 * timeFraction)
 * }
 */
export function animate(
  drawCallBack: (progress: number) => unknown,
  duration: number,
  { timing = (timeFraction: number) => timeFraction } = {}
) {
  let start = performance.now();

  requestAnimationFrame(function animate(time) {
    let timeFraction = (time - start) / duration;
    if (timeFraction > 1) timeFraction = 1;

    let progress = timing(timeFraction);

    drawCallBack(progress);

    if (timeFraction < 1) {
      requestAnimationFrame(animate);
    }
  });
}

export function power(timeFraction, {power = 2} = {}) { 
      return Math.pow(timeFraction, power)
    }
export function circ(timeFraction) {
  return 1 - Math.sin(Math.acos(timeFraction));
}
export function shotFromABow(x, timeFraction) {
  return Math.pow(timeFraction, 2) * ((x + 1) * timeFraction - x);
}
export function bounce(timeFraction) {
  for (let a = 0, b = 1; 1; a += b, b /= 2) {
    if (timeFraction >= (7 - 4 * a) / 11) {
      return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2);
    }
  }
}
export function elastic(x, timeFraction) {
  return Math.pow(2, 10 * (timeFraction - 1)) * Math.cos(((20 * Math.PI * x) / 3) * timeFraction);
}

/**
 * Jumping animation of typing text
 */
export function animateTextArea(textArea: HTMLTextAreaElement) {
  let text = textArea.value;
  let to = text.length,
    from = 0;

  animate(
    (progress) => {
      let result = (to - from) * progress + from;
      textArea.value = text.substr(0, Math.ceil(result));
    },
    5000,
    { timing: bounce }
  );

  function bounce(timeFraction: number): number {
    for (let a = 0, b = 1, result; 1; a += b, b /= 2) {
      if (timeFraction >= (7 - 4 * a) / 11) {
        return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2);
      }
    }

    return 0;
  }
}

/**
 *
 * @param timing - easeIn function
 * @returns easeOut function
 * @example
 * function bounce(timeFraction) {
 *   for (let a = 0, b = 1, result; 1; a += b, b /= 2) {
 *     if (timeFraction >= (7 - 4 * a) / 11) {
 *       return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2)
 *     }
 *   }
 * }
 *
 * let bounceEaseOut = makeEaseOut(bounce);
 */
export function makeEaseOut(timingFunction: (timeFraction: number) => number) {
  return function (timeFraction: number) {
    return 1 - timingFunction(1 - timeFraction);
  };
}
/**
 *
 * @param timing - easeIn function
 * @returns easeInOut function
 * @example
 * function bounce(timeFraction) {
 *   for (let a = 0, b = 1, result; 1; a += b, b /= 2) {
 *     if (timeFraction >= (7 - 4 * a) / 11) {
 *       return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2)
 *     }
 *   }
 * }
 *
 * let bounceEaseInOut = makeEaseInOut(bounce);
 */
export function makeEaseInOut(timingFunction: (timeFraction: number) => number) {
  return function (timeFraction: number) {
    if (timeFraction < 0.5) return timingFunction(2 * timeFraction) / 2;
    else return (2 - timingFunction(2 * (1 - timeFraction))) / 2;
  };
}
