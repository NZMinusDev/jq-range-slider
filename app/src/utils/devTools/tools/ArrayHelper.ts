/**
 * Adds missing and removes extra elements
 */
const fixLength = <TArray extends unknown[]>(
  arr: TArray,
  desiredLength: number,
  filler: TArray[number]
) => {
  const previousLength = arr.length;
  if (arr.length !== desiredLength) {
    // eslint-disable-next-line no-param-reassign
    arr.length = desiredLength;
    // eslint-disable-next-line no-param-reassign
    arr = arr.fill(filler, previousLength);
  }
};

// eslint-disable-next-line import/prefer-default-export
export { fixLength };
