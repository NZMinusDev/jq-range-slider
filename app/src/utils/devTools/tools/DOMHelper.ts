export function has(
  elements: NodeListOf<Element> | HTMLCollection | Element[],
  contained: string | Element
): Element | undefined {
  if (typeof contained === "string") {
    return Array.from(elements).find((element) => element.querySelector(contained));
  } else {
    return Array.from(elements).find((element) => element.contains(contained));
  }
}
export function hasAll(
  elements: NodeListOf<Element> | HTMLCollection | Element[],
  contained: string | Element
): Element[] | undefined {
  let isHas: Element[] = [];
  if (typeof contained === "string") {
    isHas = Array.from(elements).filter((element) => element.querySelector(contained));
  } else {
    isHas = Array.from(elements).filter((element) => element.contains(contained));
  }

  return isHas.length > 0 ? isHas : undefined;
}

/** Cross-browsers document.documentElement.scrollHeight/scrollWidth */
export function getDocumentSizes() {
  return {
    scrollHeight: Math.max(
      document.body.scrollHeight,
      document.documentElement.scrollHeight,
      document.body.offsetHeight,
      document.documentElement.offsetHeight,
      document.body.clientHeight,
      document.documentElement.clientHeight
    ),
    scrollWidth: Math.max(
      document.body.scrollHeight,
      document.documentElement.scrollWidth,
      document.body.offsetHeight,
      document.documentElement.offsetWidth,
      document.body.clientHeight,
      document.documentElement.clientWidth
    ),
  };
}
export function getWindowSizes() {
  return {
    width: document.documentElement.clientWidth, // window doesn't take into account scrollbar
    height: document.documentElement.clientHeight,
  };
}
export function isHidden(elem: HTMLElement) {
  return !elem.offsetWidth && !elem.offsetHeight;
}
/** @see module:PluginCreation/eventMixin */
export function setFullHeight(element: HTMLElement) {
  element.style.height = `${element.scrollHeight}px`;
}
enum Direction {
  Up = "Up",
  Down = "Down",
  Left = "Left",
  Right = "Right",
}
/**
 * Cross-browsers(be aware of https://bugs.webkit.org/show_bug.cgi?id=5991 and use document.body instead of document.documentElement for old browsers only) extended scrollTo
 * @see module:PluginCreation/eventMixin
 */
export function scrollTo(
  element: HTMLElement,
  pageX: number | Direction.Left | Direction.Right,
  pageY: number | Direction.Up | Direction.Down
) {
  if (typeof pageX === "number") {
    element.scrollLeft = pageX;
  } else {
    switch (pageX) {
      case Direction.Left: {
        element.scrollLeft = 0;
        break;
      }
      case Direction.Right: {
        element.scrollLeft = Infinity;
        break;
      }
    }
  }
  if (typeof pageY === "number") {
    element.scrollTop = pageY;
  } else {
    switch (pageY) {
      case Direction.Up: {
        element.scrollTop = 0;
        break;
      }
      case Direction.Down: {
        element.scrollTop = Infinity;
        break;
      }
    }
  }
}
/**
 * Cross-browsers(be aware of https://bugs.webkit.org/show_bug.cgi?id=5991 and use document.body instead of document.documentElement for old browsers only) scrollBy
 * @see module:PluginCreation/eventMixin
 */
export function scrollBy(element: HTMLElement, x: number, y: number) {
  element.scrollLeft += x;
  element.scrollTop += y;
}
/** @see module:PluginCreation/eventMixin */
export function freezeScroll(element: HTMLElement, { freeze = true } = {}) {
  if (freeze === true) {
    const clientWidthBeforeFreeze = element.clientWidth;
    element.style.overflow = "hidden";
    element.style.padding += element.clientWidth - clientWidthBeforeFreeze; // vanished scrollbar size
  } else {
    element.style.overflow = "";
  }
}
/**
 * @returns proper css coordinates for position:absolute
 * @see module:PluginCreation/eventMixin
 * */
export function getCoordinatesOfAbsoluteElement(element: HTMLElement) {
  const DOMRect = element.getBoundingClientRect();

  return {
    top: DOMRect.top + pageYOffset,
    right: document.documentElement.clientWidth - DOMRect.right + pageYOffset,
    left: DOMRect.left + pageXOffset,
    bottom: document.documentElement.clientHeight - DOMRect.bottom + pageXOffset,
  };
}
/** @see module:PluginCreation/eventMixin */
export function place(element: HTMLElement, { position = "absolute" } = {}) {
  element.style.position = position;

  let coords =
    position === "absolute"
      ? getCoordinatesOfAbsoluteElement(element)
      : element.getBoundingClientRect(); // fixed

  element.style.left = coords.left + "px";
  element.style.top = coords.bottom + "px";
}

/**
 *
 * @param src
 * @param callback
 * @example
 * // example 1
 * let promise = loadScript("https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.11/lodash.js");
 *
 * promise.then(
 *  script => alert(`${script.src} loaded!`),
 *  error => alert(`Error: ${error.message}`)
 * );
 *
 * promise.then(script => alert('Another handler...'));
 *
 * // example 2
 * loadScript("/article/promise-chaining/one.js")
 *  .then(script => loadScript("/article/promise-chaining/two.js"))
 *  .then(script => loadScript("/article/promise-chaining/three.js"))
 *  .then(script => {
 *    // the scripts are loaded, we can use the functions declared in them
 *    one();
 *    two();
 *    three();
 * });
 *
 */
export function loadScript(src: URL) {
  return new Promise(function (resolve, reject) {
    let script = document.createElement("script");
    script.src = src.toString();

    script.onload = () => resolve(script);
    script.onerror = () => reject(new Error(`Script loading error ${src}`));

    document.head.append(script);
  });
}
