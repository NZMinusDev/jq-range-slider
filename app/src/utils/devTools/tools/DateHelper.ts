/**
 * @returns milliseconds
 */
export function diffDate(date1: Date, date2: Date) {
  return date2.getTime() - date1.getTime();
}
