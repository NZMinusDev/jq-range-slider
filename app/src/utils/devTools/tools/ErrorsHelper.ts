export class MyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = this.constructor.name;
  }
}

export class ErrorWrapper extends Error {
  public cause: Error;

  constructor(message: string, cause: Error) {
    super(message);
    this.cause = cause;
    this.name = "ReadError";
  }
}

/**
 * 
 * @example
 * type NetworkLoadingState = { state: "loading"; };
 * type NetworkFailedState = { state: "failed"; code: number; };
 * type NetworkSuccessState = {
 *   state: "success";
 *   response: {
 *    title: string;
 *    duration: number;
 *    summary: string;
 *   };
 * };
 * 
 * // Create a type which represents only one of the above types
 * // but you aren't sure which it is yet.
 * type NetworkState =
 *   | NetworkLoadingState
 *   | NetworkFailedState
 *   | NetworkSuccessState;
 * 
 * function logger(state: NetworkState): string {
 *   // Right now TypeScript does not know which of the three
 *   // potential types state could be.
 *   // Trying to access a property which isn't shared
 *   // across all types will raise an error
 *   state.code;
 *   // By switching on state, TypeScript can narrow the union
 *   // down in code flow analysis
 *   switch (state.state) {
 *     case "loading":
 *       return "Downloading...";
 *     case "failed":
 *       // The type must be NetworkFailedState here,
 *       // so accessing the `code` field is safe
 *       return `Error ${state.code} downloading`;
 *     case "success":
 *       return `Downloaded ${state.response.title} - ${state.response.summary}`;
 *     default:
 *       return assertNever(s);
 *   }
 * }
 */
export function assertNever(valueToCheck: never): never {
  throw new Error("Unexpected object: " + valueToCheck);
}