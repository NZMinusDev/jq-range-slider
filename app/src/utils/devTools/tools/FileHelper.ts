export function createDataURL(
  blobParts: BlobPart[],
  options: BlobPropertyBag = { type: "text/plain" }
) {
  let blob = new Blob(blobParts, options);
  return URL.createObjectURL(blob);
}

export async function createBase64(
  blobParts: BlobPart[],
  options: BlobPropertyBag = { type: "text/plain" }
) {
  let blob = new Blob(blobParts, options);

  let fileReader = new FileReader();
  fileReader.readAsDataURL(blob);

  return await new Promise<string>((resolve) => {
    fileReader.onload = function () {
      resolve(fileReader.result as string);
    };
  });
}

export async function blobToArrayBuffer(blob: Blob) {
  let fileReader = new FileReader();
  fileReader.readAsArrayBuffer(blob);

  return await new Promise<ArrayBuffer>((resolve) => {
    fileReader.onload = function () {
      resolve(fileReader.result as ArrayBuffer);
    };
  });
}

export function imgToCanvas(
  img: HTMLImageElement,
  imgProcessing = function (canvas: HTMLCanvasElement) {}
) {
  let canvas = document.createElement("canvas");
  canvas.width = img.clientWidth;
  canvas.height = img.clientHeight;

  let context = canvas.getContext("2d");
  context?.drawImage(img, 0, 0);

  imgProcessing(canvas);

  return canvas;
}
export async function imgToBlob(
  img: HTMLImageElement,
  imgProcessing = function (canvas: HTMLCanvasElement) {}
) {
  const canvas = imgToCanvas(img, imgProcessing);

  return await new Promise<Blob>((resolve) => canvas.toBlob(resolve as BlobCallback, "image/png"));
}
