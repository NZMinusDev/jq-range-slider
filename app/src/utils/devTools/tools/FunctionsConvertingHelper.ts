import { GenericFunc } from "./TypingHelper";

/**
 * Currying of function
 * @param func
 * @example
 * function sum(a, b, c) {
 *  return a + b + c;
 *}
 *
 *let curriedSum = curry(sum);
 *
 *alert( curriedSum(1, 2, 3) ); // 6
 *alert( curriedSum(1)(2,3) ); // 6
 *alert( curriedSum(1)(2)(3) ); // 6
 */
export function curry(func: (...args: unknown[]) => unknown) {
  return function curried(...args: unknown[]) {
    if (args.length >= func.length) {
      return func.apply(this, args);
    } else {
      return function (...args: unknown[]) {
        return curried.apply(this, args.concat(args));
      };
    }
  };
}

/**
 * Clone function
 * @param func function to clone
 * @returns clone of function
 * @example
 * const test1 =  function(a,b,c) {
 *   return a+b+c;
 * };
 *
 * console.log(test1 === clone(test1)); // false
 * console.log(test1(1,1,1) === clone(test1)(1,1,1)); // true
 * console.log(clone(test1)(1,1,1)); // 3
 */
export function cloneFunction<TFuncArgs extends unknown[], TFuncReturn>(
  func: GenericFunc<TFuncArgs, TFuncReturn> | any
) {
  const that = func;
  const temp = function () {
    return that.apply(func, (arguments as unknown) as TFuncArgs);
  };

  for (let key in func) {
    if (func.hasOwnProperty(key)) {
      temp[key] = func[key];
    }
  }

  Object.defineProperty(temp, "name", { value: func.name });

  return temp;
}
