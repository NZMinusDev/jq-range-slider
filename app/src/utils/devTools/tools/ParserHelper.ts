export function parseFileExtension(fullPath: string) {
  const filePath = fullPath.substring(0, fullPath.lastIndexOf(".")) || fullPath;
  const fileExt = fullPath.substring(fullPath.lastIndexOf(".") + 1, fullPath.length) || fullPath;
  return { filePath, fileExt };
}

export function getPrecision(number: number) {
  return number.toString().includes(".") ? number.toString().split(".").pop()!.length : 0;
}
