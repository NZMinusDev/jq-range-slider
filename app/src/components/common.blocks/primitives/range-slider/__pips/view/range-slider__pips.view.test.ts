import RangeSliderPipsView, { DEFAULT_OPTIONS } from "./range-slider__pips.view";

import {
  InstancePropsExpecter,
  testInit,
  testInitDEFAULT_OPTIONS,
  testGetter,
  testSetter,
  DifferentArguments,
  testDOM,
} from "@utils/devTools/tools/UnitTestingHelper";
import { collapsingParseInt } from "@utils/devTools/tools/ParserHelper";

const viewPropertiesExpecter: InstancePropsExpecter<
  ConstructorParameters<typeof RangeSliderPipsView>,
  RangeSliderPipsView
> = function ({ instance }) {
  instance["_options"].values.forEach((value, index, self) => {
    if (index > 0) expect(value.percent > self[index - 1].percent);
    expect(value.percent).toBeGreaterThanOrEqual(0);
    expect(value.percent).toBeLessThanOrEqual(100);
  });

  expect(instance["_options"].density).toBeGreaterThanOrEqual(0);
  expect(instance["_options"].density).toEqual(
    collapsingParseInt(`${instance["_options"].density}`)
  );
};

const differentOptionsArg: DifferentArguments<Parameters<
  typeof RangeSliderPipsView.prototype.setOptions
>> = {
  invalidOptionalArguments: [
    [
      {
        values: [
          { value: 0, percent: 9 },
          { value: -100, percent: 81 },
          { value: 100, percent: 100 },
          { value: 20, percent: 92.72 },
          { value: -1000, percent: 0 },
        ],
      },
    ],
    [
      {
        values: [
          { value: 0, percent: -10 },
          { value: 100, percent: 101 },
        ],
      },
    ],
    [{ density: -5 }],
    [{ density: 5.5543 }],
  ],
  partialOptionalArguments: [
    [
      {
        values: [
          { value: -100, percent: 0 },
          { value: 0, percent: 50 },
          { value: 100, percent: 100 },
        ],
      },
    ],
  ],
  fullOptionalArguments: [
    [
      {
        orientation: "horizontal",
        isHidden: true,
        values: [
          { value: -200, percent: 0 },
          { value: 0, percent: 66.66 },
          { value: 10, percent: 70 },
          { value: 100, percent: 100 },
        ],
        density: 2,
        formatter: (value: number) => `${value.toString()}$`,
      },
    ],
  ],
};

testInitDEFAULT_OPTIONS(RangeSliderPipsView, [DEFAULT_OPTIONS], viewPropertiesExpecter);

testInit({
  Creator: RangeSliderPipsView,
  differentConstructorArgs: {
    validRequiredArguments: [[]],
    ...(differentOptionsArg as DifferentArguments<
      ConstructorParameters<typeof RangeSliderPipsView>
    >),
  },
  instancePropsExpecter: viewPropertiesExpecter,
  propsToSet: new Map().set("_options", 1),
});
describe("init", () => {
  describe("with default options", () => {
    test("the instance's func options should be to have returned", () => {
      const instance = new RangeSliderPipsView();

      const formatterMock = jest.fn(instance["_options"].formatter);
      formatterMock(1);
      expect(formatterMock).toHaveReturned();

      const templateMock = jest.fn(instance.template);
      templateMock();
      expect(templateMock).toHaveReturned();
    });
  });
});

testGetter({
  Creator: RangeSliderPipsView,
  constructorArgs: [],
  instancePropsExpecter: viewPropertiesExpecter,
  methodOfInstanceToTest: {
    methodReference: RangeSliderPipsView.prototype.getOptions,
    expecter: ({ mock, passedArgs, instance }) => {},
    returns: "_options",
  },
});
testSetter({
  Creator: RangeSliderPipsView,
  constructorArgs: [],
  instancePropsExpecter: viewPropertiesExpecter,
  methodOfInstanceToTest: {
    methodReference: RangeSliderPipsView.prototype.setOptions,
    expecter: ({ mock, passedArgs, instance }) => {},
    differentArguments: differentOptionsArg,
  },
  propsToSet: new Map().set("_options", 0),
  resetPropsTo: new Map().set("_options", DEFAULT_OPTIONS),
});

testDOM({
  Creator: RangeSliderPipsView,
  constructorsArgs: [],
  templatesArgs: [],
  callbacksWithTest: [],
});
