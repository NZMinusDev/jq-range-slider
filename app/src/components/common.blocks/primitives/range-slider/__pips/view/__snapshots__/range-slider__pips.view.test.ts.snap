// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOM manipulation renders correctly 1`] = `
<DocumentFragment>
  <!---->
  <div
    class="range-slider__pips range-slider__pips_orientation-horizontal"
    style=""
  >
    
      
    
    
  </div>
  <!---->
</DocumentFragment>
`;
